python-package/lightgbm/basic.py

```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+                    'left_child': None,
+                    'right_child': None,
+                    'parent_index': parent_node,
+                    'split_feature': _get_split_feature(t, feature_names),
+                    'split_gain': None,
+                    'threshold': None,
+                    'decision_type': None,
+                    'missing_direction': None,
+                    'missing_type': None,
+                    'leaf_weight': None,
+                    'internal_count': None}
+
+            # Update values to reflect node type (leaf or split)
+            if _is_split_node(t):
+                node['left_child'] = _get_node_index(t['left_child'], tree_index)
+                node['right_child'] = _get_node_index(t['right_child'], tree_index)
+                node['split_gain'] = t['split_gain']
+                node['threshold'] = t['threshold']
+                node['decision_type'] = t['decision_type']
+                node['missing_direction'] = 'left' if t['default_left'] else 'right'
+                node['missing_type'] = t['missing_type']
+                node['internal_count'] = t['internal_count']
+            else:
+                node['leaf_weight'] = t['leaf_value']
+                node['internal_count'] = t['leaf_count']
+
+            return node
+
+        def tree_dict_to_node_list(t, node_depth=1, tree_index=None,
+                                   feature_names=None, parent_node=None):
+
+            node = create_node_record(t,
+                                      node_depth=node_depth,
+                                      tree_index=tree_index,
+                                      feature_names=feature_names,
+                                      parent_node=parent_node)
+
+            res = [node]
+
+            if _is_split_node(t):
+                # traverse the next level of the tree
+                children = ['left_child', 'right_child']
+                for child in children:
+                    subtree_list = tree_dict_to_node_list(
+                        t[child],
+                        node_depth=node_depth + 1,
+                        tree_index=tree_index,
+                        feature_names=feature_names,
+                        parent_node=node['node_index'])
+                    # In tree format, "subtree_list"is a list of node records (dicts), and we
+                    # add node to the list.
+                    res.extend(subtree_list)
+            return res
+
+        m_dict = self.dump_model()
+        feature_names = m_dict['feature_names']
+        m_list = []
+        for t in m_dict['tree_info']:
+            m_list.extend(tree_dict_to_node_list(t['tree_structure'],
+                                                 tree_index=t['tree_index'],
+                                                 feature_names=feature_names))
+
+        if PANDAS_INSTALLED:
+            return DataFrame(m_list, columns=m_list[0].keys())
```

Comment:
`m_list[0].keys()` is not expected to be sorted and it can cause problems in different Python versions/implementations:
> Changed in version 0.25.0: If data is a list of dicts, column order follows insertion-order for Python 3.6 and later.
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html

I guess you can simply omit `columns` argument and pandas will name columns properly by default.

Reply:
I don't understand this suggestion. Not passing in the columns from `m_list[0].keys()` will cause the columns to be sorted alphabetically in Pandas versions prior to 0.25.0. Therefore, to ensure consistency in all versions of Pandas, we explicitly pass the order of the columns.

Reply:
Ah, I see! But now there is no particular order at all, as `m_list[0].keys()` is a set-like collection. Maybe it should be `sorted(m_list[0].keys())` to ensure that columns are always sorted alphabetically for any Python implementation and pandas version?

Reply:
My intention is not to have the columns be alphabetically sorted even thought that would be consistent across versions of python using `sorted()` as you suggest. I think the ordering of the columns is intuitive to how one might think of a tree's hierarchy: `tree_index`, `node_index`, ...,leaf `count`. 

A coupe options: 
1.) For users with a pre-3.6 version of Python, their column ordering will be non-deterministic. I'll defer to you as a library maintainer to judge whether that is too unappealing. 
2.) Instead of using a `dict` , we use a `collections.OrderedDict` to preserve the order of the key,value pairs. Not sure if this will have any performance or other downstream effects, but it should make sure the ordering is consistent. 
3.) I hard-code the order of the columns with a non-dynamic list. I don't love this option because if dictionary keys in the `model_list` ever change, then we'll have to remember to update this static list of column names/dictionary keys.

Reply:
Yes, I love your way of ordering too. But we should be as much consistent as it possible. So, I personally see only №2 as a good workaround. Can we try it? I don't think that there will be any significant overhead in comparison to simple dict. If pandas is compatible with OrderedDict, then everything should be OK, I hope.

Reply:
OK. I went with option #2. I'm import `OrderedDict()` at the top of basic.py. I hope that's OK.